src/wayland/dmabuf/mod.rs

```diff
@@ -11,471 +11,470 @@
-
-use slog::{o, trace};
+use wayland_server::{backend::GlobalId, Client, Display, DisplayHandle, GlobalDispatch, Resource, WEnum};
 
-use crate::backend::allocator::{
-    dmabuf::{Dmabuf, DmabufFlags, Plane},
-    Format, Fourcc, Modifier,
+use crate::{
+    backend::allocator::{
+        dmabuf::{Dmabuf, DmabufFlags, Plane},
+        Format, Fourcc, Modifier,
+    },
+    utils::{ids::id_gen, UnmanagedResource},
 };
 
-const DMABUF_VERSION: u32 = 3;
+use super::buffer::{Buffer, BufferHandler};
 
-/// Initialize a dmabuf global.
+/// Delegate type for all dmabuf globals.
 ///
-/// You need to provide a vector of the supported formats, as well as a closure,
-/// that will validate the parameters provided by the client and tests the import as a dmabuf.
-pub fn init_dmabuf_global<F, L>(
-    display: &mut Display,
-    formats: Vec<Format>,
-    handler: F,
-    logger: L,
-) -> Global<zwp_linux_dmabuf_v1::ZwpLinuxDmabufV1>
-where
-    L: Into<Option<::slog::Logger>>,
-    F: for<'a> FnMut(&Dmabuf, DispatchData<'a>) -> bool + 'static,
-{
-    display.create_global(DMABUF_VERSION, dmabuf_global(formats, handler, logger))
+/// Dmabuf globals are created using this type and events will be forwarded to an instance of the dmabuf global.
+#[derive(Debug)]
+pub struct DmabufState {
+    /// Globals managed by the dmabuf handler.
+    globals: HashMap<usize, GlobalId>,
+}
+
+impl DmabufState {
+    /// Creates a new [`DmabufState`] delegate type.
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> DmabufState {
+        DmabufState {
+            globals: HashMap::new(),
+        }
+    }
+
+    /// Creates a dmabuf global with the specified supported formats.
+    pub fn create_global<D, L>(
+        &mut self,
+        display: &mut Display<D>,
+        formats: Vec<Format>,
+        logger: L,
+    ) -> DmabufGlobal
+    where
+        D: GlobalDispatch<zwp_linux_dmabuf_v1::ZwpLinuxDmabufV1, GlobalData = DmabufGlobalData>
+            + BufferHandler
+            + DmabufHandler
+            + 'static,
+        L: Into<Option<::slog::Logger>>,
+    {
+        self.create_global_with_filter(display, formats, |_| true, logger)
+    }
+
+    /// Creates a dmabuf global with the specified supported formats.
+    ///
+    /// This function unlike [`DmabufState::create_global`] also allows you to specify a filter function to
+    /// determine which clients may see this global. This functionality may be used on multi-gpu systems in
+    /// order to make a client choose the correct gpu.
+    pub fn create_global_with_filter<D, F, L>(
+        &mut self,
+        display: &mut Display<D>,
+        formats: Vec<Format>,
+        filter: F,
+        logger: L,
+    ) -> DmabufGlobal
+    where
+        D: GlobalDispatch<zwp_linux_dmabuf_v1::ZwpLinuxDmabufV1, GlobalData = DmabufGlobalData>
+            + BufferHandler
+            + DmabufHandler
+            + 'static,
+        F: for<'c> FnMut(&'c Client) -> bool + Send + 'static,
```

Comment:
Do we need `FnMut` here? I'm under the impression that most filter implementations woudl not need to have mutable internal state, so `Fn` would be enought. That'd allow to remove the inner `Mutex<_>`, and the user could still put one in the closure if they really need mutable state in the filter.